### DISCUSSION TITLE
Camera mounted far from meter

### DISCUSSION THREAD

**Discussion by jasaw** [jomjol/AI-on-the-edge-device#2733] [Ideas]
At places where there are regulations on the minimum separation between a gas meter and a device (ESP32-CAM) that is not intrinsic safety certified. The ESP32-CAM would need to be mounted 1 meter to 3 meters away from the gas meter. Would it be possible to take the photo at higher resolution like UXGA (1600 x 1200) and crop the part that has the meter and down sample it to VGA? This step would be done before aligning the photo step.

If we don't have enough memory to handle higher resolution photo, could we delete the TFLite object from memory to make space for higher resolution photo, and reinstantiate and reload the model after photo has been down sampled?

Alternatively, we keep the VGA resolution photo but take the RGB (`pixel_format = PIXFORMAT_RGB565`) format of photo, crop it, then only JPEG encode the cropped photo to minimize noise. The cropping works on existing framebuffer memory, so no additional memory required. Pseudo-code below.

```
void crop_image(camera_fb_t *fb, unsigned short cropLeft, unsigned short cropRight, unsigned short cropTop, unsigned short cropBottom)
{
    unsigned int maxTopIndex = cropTop * fb->width * 2;
    unsigned int minBottomIndex = ((fb->width*fb->height) - (cropBottom * fb->width)) * 2;
    unsigned short maxX = fb->width - cropRight; // In pixels
    unsigned short newWidth = fb->width - cropLeft - cropRight;
    unsigned short newHeight = fb->height - cropTop - cropBottom;
    size_t newJpgSize = newWidth * newHeight *2;

    unsigned int writeIndex = 0;
    // Loop over all bytes
    for(int i = 0; i < fb->len; i+=2){
        // Calculate current X, Y pixel position
        int x = (i/2) % fb->width;

        // Crop from the top
        if(i < maxTopIndex){ continue; }

        // Crop from the bottom
        if(i > minBottomIndex){ continue; }

        // Crop from the left
        if(x <= cropLeft){ continue; }

        // Crop from the right
        if(x > maxX){ continue; }

        // If we get here, keep the pixels
        fb->buf[writeIndex++] = fb->buf[i];
        fb->buf[writeIndex++] = fb->buf[i+1];
    }

    // Set the new dimensions of the framebuffer for further use.
    fb->width = newWidth;
    fb->height = newHeight;
    fb->len = newJpgSize;
}


camera_fb_t *fb = esp_camera_fb_get();
crop_image(fb, 150, 150, 50, 50);

// Create a buffer for the JPG in psram
uint8_t * jpg_buf = (uint8_t *) heap_caps_malloc(200000, MALLOC_CAP_SPIRAM | MALLOC_CAP_8BIT);

if (jpg_buf != NULL) {
    size_t jpg_size = 0;
    // Convert the RAW image into JPG
    fmt2jpg(fb->buf, fb->len, fb->width, fb->height, fb->format, 31, &jpg_buf, &jpg_size);
}
```

I'm not familiar with the code base, so I would appreciate it if someone could point me to the places in code to modify to support VGA resolution RGB format then crop and encode option.

**Comment by jomjol** [comment]
Taking larger images is limited by the RAM of the ESP32.  The crop method might be possible, but currently nobody is working on this. 
We are happy if you could make a suggestion.

**Comment by jasaw** [comment]
@jomjol I've been experimenting with cropping the image on the camera sensor itself. The ov2640 and ov5640 are capable of such function. I've been mainly playing with ov2640 and I managed to get it to work, therefore no additional RAM needed.

I maxed out the `frame_size` in `camera_config` to `FRAMESIZE_UXGA` for ov2640, then call `s->set_res_raw` function to set the resolution that I want to crop. The cropping resolution choice is very limited because of sensor hardware limitation, so it's 800x600. The `s->set_res_raw` function also sets the output resolution. The cropped 800x600 image appears to be resized to 640x480. If output resolution is 320x240, it appears to be resized to 640x480 then cropped to 320x240.

I've also introduced a few more parameters in my `Config.ini` file to allow me to quickly play with the cropping and panning without recompiling, and I pass those parameters into my `SetQualitySize` function below.

Would this feature be of interest to this project?


Changes that I've done on `ClassControllCamera.cpp`:
```C++
static camera_config_t camera_config = {
    .pin_pwdn = CAM_PIN_PWDN,
    .pin_reset = CAM_PIN_RESET,
    .pin_xclk = CAM_PIN_XCLK,
    .pin_sscb_sda = CAM_PIN_SIOD,
    .pin_sscb_scl = CAM_PIN_SIOC,

    .pin_d7 = CAM_PIN_D7,
    .pin_d6 = CAM_PIN_D6,
    .pin_d5 = CAM_PIN_D5,
    .pin_d4 = CAM_PIN_D4,
    .pin_d3 = CAM_PIN_D3,
    .pin_d2 = CAM_PIN_D2,
    .pin_d1 = CAM_PIN_D1,
    .pin_d0 = CAM_PIN_D0,
    .pin_vsync = CAM_PIN_VSYNC,
    .pin_href = CAM_PIN_HREF,
    .pin_pclk = CAM_PIN_PCLK,

    //XCLK 20MHz or 10MHz for OV2640 double FPS (Experimental)
    .xclk_freq_hz = 20000000,             // Orginal value
//    .xclk_freq_hz =    5000000,         // Test to get rid of the image errors !!!! Hangs in version 9.2 !!!!
    .ledc_timer = LEDC_TIMER_0,
    .ledc_channel = LEDC_CHANNEL_0,

    .pixel_format = PIXFORMAT_JPEG, //YUV422,GRAYSCALE,RGB565,JPEG
    // .frame_size = FRAMESIZE_VGA,    //QQVGA-UXGA Do not use sizes above QVGA when not JPEG
    .frame_size = FRAMESIZE_UXGA,    //QQVGA-UXGA Do not use sizes above QVGA when not JPEG
    .jpeg_quality = 4, //0-63 lower number means higher quality
    .fb_count = 1,       //if more than one, i2s runs in continuous mode. Use only with JPEG
    .fb_location = CAMERA_FB_IN_PSRAM, /*!< The location where the frame buffer will be allocated */
    .grab_mode = CAMERA_GRAB_LATEST,      // only from new esp32cam version
};

/*
* resolution = 0 \\ 1600 x 1200
* resolution = 1 \\  800 x  600
* resolution = 2 \\  400 x  296
*/
void CCamera::SetCamWindow(sensor_t *s, int resolution, int xOffset, int yOffset, int xLength, int yLength)
{
    s->set_res_raw(s, resolution, 0, 0, 0, xOffset, yOffset, xLength, yLength, xLength, yLength, false, false);
}


void CCamera::SetQualitySize(int qual, framesize_t resol, int zoom, int zoomOffsetX, int zoomOffsetY, bool grayscale)
{
    qual = min(63, max(8, qual)); // Limit quality from 8..63 (values lower than 8 tent to be unstable)
    
    ActualResolution = resol;
    ActualQuality = qual;

    imageZoom = zoom;
    imageZoomOffsetX = zoomOffsetX;
    imageZoomOffsetY = zoomOffsetY;
    imageGrayscale = grayscale;

    if (resol == FRAMESIZE_QVGA)
    {
        image_height = 240;
        image_width = 320;
    }
    else if (resol == FRAMESIZE_VGA)
    {
        image_height = 480;
        image_width = 640;
    }

    sensor_t * s = esp_camera_sensor_get();
    if (s) {
        // s->set_quality(s, qual);
        // s->set_framesize(s, resol);
        if (grayscale) {
            s->set_special_effect(s, 2); // 0 to 6 (0 - No Effect, 1 - Negative, 2 - Grayscale, 3 - Red Tint, 4 - Green Tint, 5 - Blue Tint, 6 - Sepia)
        } else {
            s->set_special_effect(s, 0);
        }
        LogFile.WriteToFile(ESP_LOG_ERROR, TAG, "set_res_raw: " + std::to_string(image_width) + "x" + std::to_string(image_height));
        if (zoom > 1)
            zoom = 1;
        if (image_width >= 800 || image_height >= 600) {
            zoom = 0;
        }
        if (zoom == 1) {
            if (zoomOffsetX > (800 - image_width))
                zoomOffsetX = 800 - image_width;
            if (zoomOffsetY > (600 - image_height))
                zoomOffsetY = 600 - image_height;
        } else {
            if (zoomOffsetX > (1600 - image_width))
                zoomOffsetX = 1600 - image_width;
            if (zoomOffsetY > (1200 - image_height))
                zoomOffsetY = 1200 - image_height;
        }
        SetCamWindow(s, zoom, zoomOffsetX, zoomOffsetY, image_width, image_height);
    }
    else {
        LogFile.WriteToFile(ESP_LOG_ERROR, TAG, "SetQualitySize: Failed to get control structure");
    }
}
```

**Comment by jasaw** [comment]
@jomjol I've played around with the cropping on sensor a bit more and managed to get my OV2640 to take 1600x1200 resolution cropped to 640x480, or 800x600 cropped to 640x480 (or 320x240 depends on config). I've also made grayscale mode to work together with brightness/contrast/saturation control. I've also added support for adjusting auto exposure level, which is very important for certain lighting conditions.
I've put my work here for your reference.
https://github.com/jasaw/AI-on-the-edge-device/tree/feature/crop_image


![gasmeter_night_exposure_negative](https://github.com/jomjol/AI-on-the-edge-device/assets/721280/b8e0ea85-b21b-436e-a80c-cd7958d7a5d6)

**Reply by spanzetta** [reply]
Hi @jasaw 
can you pls guide me on how to install this "update" that enables all these new settings (zoom, auto exposure, grayscale)?
I tried already by copying the SD card contents into my setup.. but I still see the orginal interface.. 
I am sure I am doing something wrong.. so if you can me some tips I will try
Thanks in advance

**Reply by jasaw** [reply]
@spanzetta Just for trying out this feature, you'll need to start with a standard unmodified installation (which you already have), and you'll have to download both the ZIP files here, extract them, and
1. Install the `firmware.bin` file via the `OTA Update` feature through the web interface, or use `esptool.py` to write it straight to flash.
2. Replace the `edit_reference.html` and `readconfigparam.js` in the sd-card html directory. This can be done via the web interface file server (delete the original files first then upload new files) or copy via an SD card reader on a computer. After you've overwritten these 2 files, you'll need to force refresh the web interface on your web browser, or clear web browser cache.

Beware that the web interface `settings --> configuration` page does not expose the new `TakeImage` parameters yet, so saving a new config via the web interface configuration page will remove the new `TakeImage` parameters. If that happens, just recreate and save a new reference image (under `settings --> alignment --> reference image`). When I get some free time, I'll try to add the new `TakeImage` parameters to the web interface configuration page.

[html.zip](https://github.com/jomjol/AI-on-the-edge-device/files/14010374/html.zip)
[firmware.zip](https://github.com/jomjol/AI-on-the-edge-device/files/14010363/firmware.zip)


Also worth pointing out:
1. Light reflection issue: Don't point your ESP32cam directly at your meter. It's better to angle it a bit so the LED on the ESP32cam does not reflect directly back to the camera. You'll get less light reflection this way. Alternatively, you could use paper or thin plastic to defuse the light a bit.
2. In certain lighting conditions (e.g. evening sun hitting the meter face directly), the meter face is too bright, the auto-exposure algorithm may give you a nice overall exposure, but the meter digits are under-exposed or over-exposed. 
3. This feature is not officially supported by AI-on-the-edge-device project yet as my PR hasn't been accepted.

**Reply by spanzetta** [reply]
Many thanks for the very clear and detailed instructions..
Will try it asap and post here my feedback. 
As for lights I am planning to add external led strip (currently designing a specific 3D model for a better light positioning) because I also have big reflection problems.
Thanks again

**Reply by spanzetta** [reply]
Hi @jasaw 
I was able to use the zoom function which really resolved my problems..
And I am also impressed about the level of detection with my poor image (the water counter is even hard to read by eyes.. it's so old that it's "consumed"..) (*)
I hope that this functionality will be included as "standard" into next builds.. it's really usefull!!
Well done

(*) See below..

![image](https://github.com/jomjol/AI-on-the-edge-device/assets/49818743/a92f88ff-946d-405a-b02b-c797a38fa4b8)

**Comment by jasaw** [comment]
Created a PR: https://github.com/jomjol/AI-on-the-edge-device/pull/2799

**Comment by nliaudat** [comment]
@jasaw Have you tested an optical zoom ?

**Reply by jasaw** [reply]
No, I have not tested with a camera with optical zoom. I've been testing with the stock standard ESP32-cam dev kit which has an OV2640 (2MP) without optical zoom, which is only good for under 1 meter distance (depending on font size on the meter). OV5640 (5MP) camera with this cropping technique will give a bit more distance. Optical zoom will definitely be required to go up to 3 meters distance.

The aim here is to make it a little bit more forgiving on install distance by having an adjustable cropping / scaling feature.
From the screenshot above, assuming output of 640x480:
- zoom disabled means 1600x1200 scaled to 640x480. This is the camera configuration of this project right now. You'll get the entire frame in your output, no "zoom".
- zoom mode 0 means 1600x1200 cropped to 640x480. This gives the maximum "zoom" of the camera.
- zoom mode 1 means 1600x1200 scaled to 800x600, then cropped to 640x480. The gives mid level "zoom".

This essentially gives us 3 "zoom" levels.

**Comment by jasaw** [comment]
I've added ability to control sharpness of the image for OV2640 camera. Valid sharpness values from -3 to +3, and setting to -4 means auto-sharpness.
Also fixed a problem where the OV2640 camera only applies "negative" effect even though both grayscale and negative are enabled.

~~@spanzetta I've attached the relevant files for you to try. Again, this is highly experimental.~~
Unfortunately, the camera doesn't appear to be stable when I mess with the sharpness. More work needs to be done.

Update: The instability appears to be caused by me playing around with the JPEG quality.
Here are the files:
[html.zip](https://github.com/jomjol/AI-on-the-edge-device/files/14094239/html.zip)
[firmware.zip](https://github.com/jomjol/AI-on-the-edge-device/files/14094245/firmware.zip)

**Comment by jasaw** [comment]
Implemented by #2885.

**Reply by whitedavidp** [reply]
I just want to thank those who posted above regarding using zoom and negative in order to get better readings. My meter is old, has poor clarity due to age and perhaps some moisture. Using negative and zoom really helped me alot. Best!
